titus-server-master/src/main/java/com/netflix/titus/master/kubernetes/controller/PodGcController.java

```diff
@@ -0,0 +1,246 @@
+                .filter(p -> isPodOnUnknownNode(p, knownNodeNames))
+                .collect(Collectors.toList());
+
+        logger.info("Attempting to GC {} pods: {} on unknown nodes", podsOnUnknownNodes.size(),
+                podsOnUnknownNodes);
+        this.podsOnUnknownNodes.set(podsOnUnknownNodes.size());
+        for (V1Pod pod : podsOnUnknownNodes) {
+            gcPod(pod);
+        }
+        logger.info("Finished pod GC on unknown nodes");
+    }
+
+    @VisibleForTesting
+    boolean isPodOnUnknownNode(V1Pod pod, Set<String> knownNodeNames) {
+        if (pod == null || pod.getSpec() == null) {
+            return false;
+        }
+
+        String nodeName = pod.getSpec().getNodeName();
+        return StringExt.isNotEmpty(nodeName) && !knownNodeNames.contains(nodeName);
+    }
+
+    /**
+     * GC pods past deletion timestamp timeout.
+     */
+    private void gcPodsPastDeletionTimestamp(List<V1Pod> pods) {
+        List<V1Pod> podsPastDeletionTimestampToGc = pods.stream()
+                .filter(this::isPodPastDeletionTimestamp)
+                .collect(Collectors.toList());
+
+        logger.info("Attempting to GC {} pods: {} past deletion timestamp", podsPastDeletionTimestampToGc.size(),
+                podsPastDeletionTimestampToGc);
+        podsPastDeletionTimestampToGcGauge.set(podsPastDeletionTimestampToGc.size());
+        for (V1Pod pod : podsPastDeletionTimestampToGc) {
+            gcPod(pod);
+        }
+        logger.info("Finished pods past deletion timestamp GC");
+    }
+
+    @VisibleForTesting
+    boolean isPodPastDeletionTimestamp(V1Pod pod) {
+        V1PodSpec spec = pod.getSpec();
+        V1ObjectMeta metadata = pod.getMetadata();
+        if (spec == null || metadata == null) {
+            return false;
+        }
+
+        Long terminationGracePeriodSeconds = spec.getTerminationGracePeriodSeconds();
+        DateTime deletionTimestamp = metadata.getDeletionTimestamp();
+        if (terminationGracePeriodSeconds == null || deletionTimestamp == null) {
+            return false;
+        }
+
+        long terminationGracePeriodMs = terminationGracePeriodSeconds * 1000L;
+        return clock.isPast(deletionTimestamp.getMillis() + terminationGracePeriodMs
+                + configuration.getPodsPastTerminationGracePeriodMs());
+    }
+
+    /**
+     * GC pods in Pending phase with a deletion timestamp.
+     */
+    private void gcPendingPodsWithDeletionTimestamp(List<V1Pod> pods) {
+        List<V1Pod> pendingPodsWithDeletionTimestampToGc = pods.stream()
+                .filter(this::isPodInPendingPhaseWithDeletionTimestamp)
+                .collect(Collectors.toList());
+
+        logger.info("Attempting to GC {} pending pods: {} with deletion timestamp", pendingPodsWithDeletionTimestampToGc.size(),
+                pendingPodsWithDeletionTimestampToGc);
+        pendingPodsWithDeletionTimestampToGcGauge.set(pendingPodsWithDeletionTimestampToGc.size());
```

Comment:
This gauge could be set to an inflated value depending on how many pods are actually GC'd successfully vs how many fail in the gcPod() function (where the exception is swallowed). Do you think it'd be better to wait until the gcPod loop to set the gauge to the correctly number or possibly introduce a separate gauge to track attempted vs actual pod GC?